L1Trigger/L1TTrackMatch/test/L1TkMETAnalyser.cc

```diff
@@ -0,0 +1,271 @@
+#include <cmath>
+#include <iostream>
+#include <map>
+#include <string>
+#include <vector>
+
+#include "CommonTools/UtilAlgos/interface/TFileService.h"
+#include "DataFormats/Candidate/interface/Candidate.h"
+#include "DataFormats/HepMCCandidate/interface/GenParticle.h"
+#include "DataFormats/L1TCorrelator/interface/TkEtMiss.h"
+#include "DataFormats/L1TCorrelator/interface/TkEtMissFwd.h"
+#include "DataFormats/L1TrackTrigger/interface/TTTypes.h"
+#include "DataFormats/L1Trigger/interface/EtSum.h"
+#include "DataFormats/L1Trigger/interface/Vertex.h"
+#include "DataFormats/Phase2TrackerDigi/interface/Phase2TrackerDigi.h"
+#include "DataFormats/TrackerCommon/interface/TrackerTopology.h"
+#include "FWCore/Framework/interface/one/EDAnalyzer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ServiceRegistry/interface/Service.h"
+#include "Geometry/Records/interface/TrackerDigiGeometryRecord.h"
+#include "L1Trigger/L1TTrackMatch/interface/L1TkEtMissEmuAlgo.h"
+#include "TEfficiency.h"
+#include "TGraphAsymmErrors.h"
+#include "TGraphErrors.h"
+#include "TH1F.h"
+#include "TH2F.h"
+#include "TPad.h"
+#include "TProfile.h"
+#include "TTree.h"
+
+using namespace std;
+
+class L1TkMETAnalyser : public edm::one::EDAnalyzer<> {
+public:
+  explicit L1TkMETAnalyser(const edm::ParameterSet& iConfig);
+  ~L1TkMETAnalyser() override;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  void beginJob() override;
+  void analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup) override;
+  void endJob() override;
+
+  edm::ParameterSet config;
+
+  edm::InputTag TrackMETSimInputTag;
+  edm::InputTag TrackMETEmuInputTag;
+  edm::InputTag TrackMETHWInputTag;
+
+  edm::EDGetTokenT<std::vector<l1t::TkEtMiss>> TrackMETSimToken_;
+  edm::EDGetTokenT<std::vector<l1t::EtSum>> TrackMETEmuToken_;
+  edm::EDGetTokenT<std::vector<l1t::EtSum>> TrackMETHWToken_;
+
+  float EtScale;
+  float EtphiScale;
+
+  bool available_;
+  TTree* eventTree;
+
+  std::vector<float>* m_SimMET;
+  std::vector<float>* m_EmuMET;
+  std::vector<float>* m_HwMET;
+
+  std::vector<float>* m_SimMETphi;
+  std::vector<float>* m_EmuMETphi;
+  std::vector<float>* m_HwMETphi;
+
+  std::vector<float>* m_SimNtrk;
+  std::vector<float>* m_EmuNtrk;
+  std::vector<float>* m_HwNtrk;
+
+  edm::Service<TFileService> fs_;
```

Comment:
Service handle should not be used as EDModule class member variables. It should be instantiated in the local scope of the member functions where it is used instead.